app/app_client_test.go

```diff
@@ -0,0 +1,1928 @@
+		) (*pb.GetRobotPartsResponse, error) {
+			test.That(t, in.RobotId, test.ShouldEqual, robotID)
+			return &pb.GetRobotPartsResponse{
+				Parts: []*pb.RobotPart{&pbRobotPart},
+			}, nil
+		}
+		resp, err := client.GetRobotParts(context.Background(), robotID)
+		test.That(t, err, test.ShouldBeNil)
+		test.That(t, resp, test.ShouldResemble, expectedRobotParts)
+	})
+
+	t.Run("GetRobotPart", func(t *testing.T) {
+		grpcClient.GetRobotPartFunc = func(
+			ctx context.Context, in *pb.GetRobotPartRequest, opts ...grpc.CallOption,
+		) (*pb.GetRobotPartResponse, error) {
+			test.That(t, in.Id, test.ShouldEqual, partID)
+			return &pb.GetRobotPartResponse{
+				Part:       &pbRobotPart,
+				ConfigJson: configJSON,
+			}, nil
+		}
+		part, json, err := client.GetRobotPart(context.Background(), partID)
+		test.That(t, err, test.ShouldBeNil)
+		test.That(t, json, test.ShouldEqual, configJSON)
+		test.That(t, part, test.ShouldResemble, &robotPart)
+	})
+
+	t.Run("GetRobotPartLogs", func(t *testing.T) {
+		grpcClient.GetRobotPartLogsFunc = func(
+			ctx context.Context, in *pb.GetRobotPartLogsRequest, opts ...grpc.CallOption,
+		) (*pb.GetRobotPartLogsResponse, error) {
+			test.That(t, in.Id, test.ShouldEqual, partID)
+			test.That(t, in.Filter, test.ShouldEqual, &filter)
+			test.That(t, in.PageToken, test.ShouldEqual, &pageToken)
+			test.That(t, in.Levels, test.ShouldResemble, levels)
+			test.That(t, in.Start, test.ShouldEqual, &start)
+			test.That(t, in.End, test.ShouldEqual, &end)
+			test.That(t, in.Limit, test.ShouldEqual, &limit)
+			test.That(t, in.Source, test.ShouldEqual, &source)
+			return &pb.GetRobotPartLogsResponse{
+				Logs:          []*common.LogEntry{&pbLogEntry},
+				NextPageToken: pageToken,
+			}, nil
+		}
+		logs, token, err := client.GetRobotPartLogs(context.Background(), partID, &filter, &pageToken, levels, &start, &end, &limit, &source)
+		test.That(t, err, test.ShouldBeNil)
+		test.That(t, token, test.ShouldEqual, pageToken)
+		test.That(t, logs, test.ShouldResemble, logEntries)
+	})
+
+	t.Run("TailRobotPartLogs", func(t *testing.T) {
+		ch := make(chan []*LogEntry)
+		grpcClient.TailRobotPartLogsFunc = func(
+			ctx context.Context, in *pb.TailRobotPartLogsRequest, opts ...grpc.CallOption,
+		) (pb.AppService_TailRobotPartLogsClient, error) {
+			test.That(t, in.Id, test.ShouldEqual, partID)
+			test.That(t, in.ErrorsOnly, test.ShouldEqual, errorsOnly)
+			test.That(t, in.Filter, test.ShouldEqual, &filter)
+			return &mockTailRobotPartLogsClient{
+				responses: []*pb.TailRobotPartLogsResponse{
+					{Logs: []*common.LogEntry{&pbLogEntry}},
+				},
+			}, nil
+		}
+		err := client.TailRobotPartLogs(context.Background(), partID, errorsOnly, &filter, ch)
+		test.That(t, err, test.ShouldBeNil)
+
+		// var resp [][]*LogEntry
+		// for entries := range ch {
+		// 	resp = append(resp, entries)
+		// }
+		// test.That(t, resp, test.ShouldResemble, logEntries)
```

Comment:
When I try to confirm that the channel has the correct log entries, the test seems to panic. I'm not entirely sure what to do. The test succeeds without this portion.